### ISSUE TITLE
Allow extensions/plugins to alter nav display

### ISSUE THREAD

**Issue by Andre601** [mkdocs/mkdocs#2311]
> This was originally posted on https://github.com/squidfunk/mkdocs-material/issues/2318
> See the issue for any additional info needed.

## The Idea
From what I understand does MkDocs not allow extensions or plugins to customize the look of (at least) the navigation in terms of displayed text.

It would be a interesting idea to allow extensions and/or plugins to alter this navigation to allow support for things such as custom SVG icons or images.

## Example
When you want to link to an external site and want to indicate this with the [Fontawesome `external-link-alt`](https://fontawesome.com/icons/external-link-alt) could you use Materials for MkDocs and the MaterialX emoji extension together with pymdownx's emoji extension, to have something like this:
```yaml
nav:
- 'Our Website :fontawesome-solid-link-external-alt:': 'https://example.com'
```

The result would be `Our Website` with the icon right next to it.

## Alternatives
The only alternatives I can think of is using Unicode emojis directly in the text which has its own set of flaws:

- Emojis are limited. There can be cases where no emoji matches what you need.
- Emojis may look different depending on the OS used to view the page. It could perhaps even not work (Be unsupported)

## Final words
I hope you'll take a look at this.
If that is already possible, would I like to know more details about this, as it could perhaps become a part of an existing extension/plugin for future use.

**Comment by waylan** [comment]
I'm not sure why a plugin or theme couldn't add support for this. Let's consider your proposed config:

```yaml
nav:
- 'Our Website :fontawesome-solid-link-external-alt:': 'https://example.com'
```

A plugin or theme template could loop through the nav items, use a regex to detect the FontAwesome reference, and alter the nav item appropriately. A theme could just do so while building the navigation on the page. A plugin would need to store the extra data on the nav item

By the way, Python will allow you to add a new (unknown) attribute to a Python class instance:

```python
>>> from mkdocs.structure.nav import Link
>>> link = Link('Title', 'https://example.com')
>>> link.icon
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Link' object has no attribute 'icon'
>>> link.icon = 'fontawesome-solid-link-external-alt'
>>> link.icon
'fontawesome-solid-link-external-alt'
```

The only concern is that not all instances of the class are guaranteed to have the attribute defined. But with appropriate precautions (`if hasattr(link, 'icon')...`), that shouldn't be an issue.

If you only want to suppose icons for pages (not external links), a simpler solution would be to define a icon in a page's [meta-data][1]. Then a theme template could check for it and if present, include the icon in the navigation. Something like:

```markdown
---
nav-icon: fontawesome-solid-link-external-alt
---

# Some Markdown Document...
```

I suppose you could even tie the two together. Have the plugin check for the icon being specified in the nav and then add it to the meta-data of the page (while altering the page title to remove the icon reference). Then the theme would only need to check the meta-data and not need to run a regex on the page title. And of course, the plugin would not be necessary as users could simply use the meta-data directly. Of course, some users may prefer the convenience of using the nav config and could do so with the help of the plugin.

All of the above suggestions do not require any changes to be made by MkDocs itself. And, as I have no need for the feature myself, I will not be able to provide any of my very limited time to develop any of the above solutions.


[1]: https://www.mkdocs.org/user-guide/writing-your-docs/#meta-data

**Comment by Andre601** [comment]
Unfortunately do I not have the experience to make a plugin to alter the nav section to what I had in mind and from the responses in squidfunk/mkdocs-material#2318 does it seem like that at least the theme won't be able to achieve this goal here.
Extensions also seem to only target the ToC at most and a plugin would be needed for the nav?

Could I as an alternative perhaps just use a `<i>` html tag in the nav section which would then render the icon appropriately?

**Comment by waylan** [comment]
This request was to allow plugins to alter the nav. I pointed out that that is already possible. Now you are saying you don't have the experience to make a plugin. I'm confused why you made the original request then.

In any event, I have no need for or interest in this feature myself. And I have very limited time to give to MkDocs. I certainly don't have the time to develop this feature. Nor do I want to take on the maintenance of additional features. As I am not aware of any reason why a plugin couldn't implement this feature today, there is no actionable item here and I am closing this. If/when someone does develop a plugin and in the unlikely event that they encounter some limitations in MkDocs plugin API, they can file a new issue/PR to modify the API as necessary.

**Comment by Andre601** [comment]
Did you see the linked issue?
This was the main reason of this issue being opened, as the theme maintainer said it was unlikely to do for lack of features.

To be straight honest: Your message is like "I don't have interest, I don't care, so no"
Sure, it's your project, but I'm sure community members would be happy to help out.

It's just such a negative response that makes me lose motivation here.

**Comment by waylan** [comment]
I have found that if a feature is desired by enough people to justify its inclusion, then there will inevitably be someone who both wants the feature and has the time and ability to create the plugin. However, if this is something that is only desired by a couple people then it is not worth me putting the effort into it.

And to be clear, I never said 'I don't care.' I said I don't have the time. I work as a volunteer in my spare time, which is very limited. I'm not inclined to spend that limited time on a feature that only one user will use. However, if someone develops a third party plugin and it becomes wildly popular, then we will reconsider including it as a default feature.

Finally, the claim that this is not possible in the upstream issue is incorrect, as I commented there.
